File: September/18/insertinterval.py
class Solution:
    def insert(self, intervals: list[list[int]], newInterval: list[int]) -> list[list[int]]:
        res = []
        for i, (s, e) in enumerate(intervals):
            if newInterval[1] < s:
                res.append(newInterval)
                return res + intervals[i:]
            elif newInterval[0] > e:
                res.append([s, e])
            else:
                newInterval = [min(newInterval[0], s), max(newInterval[1], e)]
        res.append(newInterval)
        return res
    
    insertOneLine=lambda A,I:__import__('functools').reduce(lambda a,x:a+[x] if I[1]<x[0] and not a.append(I) else a if I[0]>x[1] or not a.append([min(I[0],x[0]),max(I[1],x[1])]) or not I.pop() else a,[x for x in A],[ ])+([I] if I else [])


s = Solution()
print(s.insert([[1,3],[6,9]], [2,5]))          # [[1,5],[6,9]]
print(s.insert([[1,2],[3,5],[6,7],[8,10],[12,16]], [4,8]))  # [[1,2],[3,10],[12,16]]

File: 2025/September/18/insertinterval.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> result = new ArrayList<>();
        int i = 0;

        // Add intervals before newInterval
        while (i < intervals.length && intervals[i][1] < newInterval[0]) {
            result.add(intervals[i++]);
        }

        // Merge overlapping intervals with newInterval
        while (i < intervals.length && intervals[i][0] <= newInterval[1]) {
            newInterval[0] = Math.min(newInterval[0], intervals[i][0]);
            newInterval[1] = Math.max(newInterval[1], intervals[i][1]);
            i++;
        }
        result.add(newInterval);

        // Add remaining intervals
        while (i < intervals.length) {
            result.add(intervals[i++]);
        }

        return result.toArray(new int[result.size()][]);
    }
}

// Example
int[][] intervals = {{1,3},{6,9}};
int[] newInterval = {2,5};
int[][] merged = new Solution().insert(intervals, newInterval);
// merged = [[1,5],[6,9]]

